src/desdemona/query.clj

```diff
@@ -58,18 +58,24 @@
        (finally
          (in-ns (ns-name old-ns)))))))
 
+(def dsl-literal?
+  "Is this a literal in the DSL?"
+  string?)
+
 (defn ^:private dsl->logic
   "Given a DSL query, compile it to the underlying logic (miniKanren)
   expressions."
   [dsl-query]
   (m/match [dsl-query]
-    [((= ((attr lvar) :seq) value) :seq)]
-    (let [lvar (free-sym lvar)]
-      `(l/featurec ~lvar {~attr ~value}))
-
-    [((= value ((attr lvar) :seq)) :seq)]
-    (let [lvar (free-sym lvar)]
-      `(l/featurec ~lvar {~attr ~value}))
+    [(('= & terms) :seq)]
+    (let [{literals true attr-terms false} (group-by dsl-literal? terms)
+          feature (fn [value [attr lvar]]
+                    `(l/featurec ~(free-sym lvar) {~attr ~value}))]
+      (m/match [(count literals) (count attr-terms)]
+        [1 1] (feature (first literals) (first attr-terms))
+        [1 _] `(l/all ~@(map (partial feature (first literals)) attr-terms))
+        [0 _] (let [u (gensym)]
+                `(l/fresh [~u] ~@(map (partial feature u) attr-terms)))))
```

Comment:
I think it's pretty great that even though this adds clear features compared to the previous version and yet it is only 2 lines longer; the only reason it's _longer_ is because it generates the same code for old samples; if it was OK to always generate a `fresh` form, it could be just as long as possibly even shorter.